Remove use of `ServiceDescriber` in MVC

dougbu opened the pull request:
- see issue aspnet/DependencyInjection#193
- use `TryAdd()` in `MvcServices` for the `IAssemblyProvider`
- fix `TestHelper.ReplaceCallContextServiceLocationService` hack
- use `TestHelper.ReplaceCallContextServiceLocationService` in `FileResultTest`
 - needed for authorization policy discovery

also fix `FiltersTest` to use correct site name

nits:
- simplify `IApplicationEnvironment` override
- `TagHelperSampleTests` does not need `ReplaceCallContextServiceLocationService`

remaining work
- remove `TestHelper.ReplaceCallContextServiceLocationService` hack
 - e.g. fully-qualify config.json path to avoid IApplicationEnvironment use
 - e.g. define an MVC service for app base resolution to wrap `IApplicationEnvironment`
 - do something for `AddAuthorization()`

Ignore RazorInstrumentationWebsite case fix; I'll split that into another PR

dougbu commented:
/cc @HaoK @davidfowl @lodejard 

please chime in on which of the follow-up points (all toward not overriding a host-provided service) matter most.  I'll file an MVC issue about whatever you feel is urgent.

hmm, we may not even need to fix `AddAuthorization()` to remove current `AddMvc(configuration)` overload.  that might be cleanup we could do separately.

dougbu commented on src/Microsoft.AspNet.Mvc/MvcServices.cs:
```diff
@@ -26,157 +24,149 @@ namespace Microsoft.AspNet.Mvc
-            yield return describe.Transient<IAssemblyProvider, DefaultAssemblyProvider>();
+            services.AddTransient<IConfigureOptions<MvcOptions>, MvcOptionsSetup>();
+            services.AddTransient<IConfigureOptions<RazorViewEngineOptions>, RazorViewEngineOptionsSetup>();
+            services.TryAdd(ServiceDescriptor.Transient<IAssemblyProvider, DefaultAssemblyProvider>());
```
allows functional tests to override this particular MVC-specific service

dougbu commented on src/Microsoft.AspNet.Mvc/MvcServices.cs:
```diff
@@ -26,157 +24,149 @@ namespace Microsoft.AspNet.Mvc
             // DefaultHtmlGenerator is pretty much stateless but depends on Scoped services such as IUrlHelper and
-            // IActionBindingContextProvider. Therefore it too is scoped.
-            yield return describe.Transient<IHtmlGenerator, DefaultHtmlGenerator>();
+            // IActionBindingContextProvider. Therefore it too is scoped.yield return describe.Transient
```
will fix this accidental paste

dougbu commented on test/Microsoft.AspNet.Mvc.FunctionalTests/ActionResultTests.cs:
```diff
@@ -324,5 +322,10 @@ public class ActionResultTests
+
+        private static void AddServices(IServiceCollection services)
+        {
+            TestHelper.AddServices(services, "ActionResultsWebSite");
```
`nameof()` if I can

HaoK commented on test/Microsoft.AspNet.Mvc.FunctionalTests/ActionResultTests.cs:
```diff
@@ -2,30 +2,28 @@
         {
             // Arrange
-            var server = TestServer.Create(_provider, _app);
+            var server = TestServer.Create(_app, AddServices);
```
nit: do you prefer this over just doing it inline?  

```
var service = TestService.Create(_app, services => TestHelper.AddServices(services, "ActionResultsWebSite"));
```

If I was going to have a private helper, I would do this and hide everything:
```
  var server = CreateTestServer();
```

HaoK replied:
Actually I've changed my mind, this isn't a nit, PLEASE change this to `var server = CreateTestServer();`  If we ever have to change this again, it will make future fixes much easier...

HaoK replied:
Actually looks like you should just move this live to TestHelper so you don't have this helper in every functional test

```
TestHelper.CreateTestServer(_app, nameof(testClass));
```

HaoK commented on test/Microsoft.AspNet.Mvc.FunctionalTests/MvcTagHelpersTest.cs:
```diff
@@ -366,14 +363,25 @@ public class MvcTagHelpersTest
             Assert.Equal(expectedContent.Trim(), responseContent.Trim());
         }
+
+        private static void AddServices(IServiceCollection services, IServiceCollection additionalServices)
```
This looks kind of wonky...

HaoK replied:
At the very least, name this something better :)

HaoK commented on test/Microsoft.AspNet.Mvc.FunctionalTests/TestHelper.cs:
```diff
@@ -20,83 +17,108 @@ public static class TestHelper
+            var assemblyProvider = CreateAssemblyProvider(applicationWebSiteName);
+            services.AddInstance<IAssemblyProvider>(assemblyProvider);
+
+            services.AddInstance<ILoggerFactory>(new LoggerFactory());
```
Do we need this logger factory?  Is no one else adding one?

dougbu replied:
could rely on the one `AddHosting()` provides but that has a comment about being removed.  so, I'm not sure but don't think it matters much at this time.

HaoK replied:
I would just try removing this line, if all the functional tests pass, good, if not, then leave it :)

dougbu replied:
removed

HaoK commented on test/Microsoft.AspNet.Mvc.FunctionalTests/TestHelper.cs:
```diff
@@ -20,83 +17,108 @@ public static class TestHelper
-            // We only want this to happen when running in proc.
-            var assembly = Assembly.Load(new AssemblyName(siteName));
+            var serviceCollection = HostingServices.Create(
+                CallContextServiceLocator.Locator.ServiceProvider,
```
I added a new Create overload today that lets you omit the CallContextServiceLocator:

https://github.com/aspnet/Hosting/blob/dev/src/Microsoft.AspNet.Hosting/HostingServices.cs#L46

dougbu replied:
nice; will use in next iteration

HaoK commented on test/Microsoft.AspNet.Mvc.FunctionalTests/TestHelper.cs:
```diff
@@ -114,34 +136,5 @@ public void Dispose()
             }
         }
-
-        // TODO: Kill this code
```
yay finally its dead!

HaoK replied:
You can resolve this https://github.com/aspnet/Mvc/issues/1612 as part of this PR

dougbu replied:
:ok:

HaoK commented:
@dougbu regarding `AddAuthorization` needing config, you can leave `AddMvc` taking config for now.  I will make a follow up pass through all of the `AddXyz()` methods once we rationalize what we want to do,  its primarily only for binding options to configuration automatically now, which still has some issues to iron out.  

That said, `AddMvc` should probably be binding against the configuration as well, unless this is being done somewhere else:
```
services.Configure<MvcOptions>(config)
```

dougbu commented:
@HaoK w.r.t. `MvcOptions`, `AddMvc` calls `services.AddOptions()` near where it calls `services.AddAuthorization(configuration)`.  look at the bottom of [`MvcServiceCollectionExtensions`](https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNet.Mvc/MvcServiceCollectionExtensions.cs#L133).

HaoK commented:
That's not quite what I meant, `AddOptions` is registering Options, `Configure<MvcOptions>(config)` would automatically try to bind the configuration against MvcOptions.  But don't worry about it, I can do as part of a consistency pass

dougbu commented:
@HaoK updated
- got rid of more functional test hacks but still need to override `IApplicationEnvironment`

fyi my plan is to get this in _after_ aspnet/Mvc#2139 but I have not rebased on top of that PR's branch
- of course, no plans to push anything this weekend

HaoK commented on samples/MvcSample.Web/Startup.cs:
```diff
@@ -46,14 +55,10 @@ public void Configure(IApplicationBuilder app)
-                    // Setup services with a test AssemblyProvider so that only the
-                    // sample's assemblies are loaded. This prevents loading controllers from other assemblies
-                    // when the sample is used in the Functional Tests.
-                    services.AddTransient<IAssemblyProvider, TestAssemblyProvider<Startup>>();
```
Nice!

HaoK commented on test/Microsoft.AspNet.Mvc.FunctionalTests/ActionResultTests.cs:
```diff
@@ -2,30 +2,27 @@
         {
             // Arrange
-            var server = TestServer.Create(_provider, _app);
+            var server = TestHelper.CreateServer(_app, SiteName);
```
One other thing that might clean things up dramatically, is it possible to move these two duplicated lines into a constructor of all the functional tests?  This would save a lot of duplication, not sure if that would work, but worth a shot...
```
          public ActionResultTests() {
             _server = TestHelper.CreateServer(_app, SiteName);
             _client = server.CreateClient();
          }
```

dougbu replied:
might work but means tests can affect each other with some xUnit runner settings.  leaving this as-is.

HaoK replied:
Well my assumption was that each xunit test is run with its own instance of the test class always, I would hope there aren't any test settings that affect that, are you sure?

See https://xunit.codeplex.com/wikipage?title=WhyDidWeBuildXunit 

"Single Object Instance per Test Method. Much has been written about why this improves test isolation. In xUnit.net we create a new instance per test"

dougbu replied:
I'm aware of that intent in xUnit but thought I'd seen issues w/ `-parallel all`. may be mixing in nUnit memories.

but why are we making this large & manual change at this time?

HaoK replied:
Feel free to file an issue to track, just seems worth doing since it would probably eliminate hundreds of lines of code

dougbu replied:
#2140

dougbu commented:
@HaoK one line removed in latest iteration

HaoK commented:
looks good to me :shipit:

HaoK commented:
@dougbu I almost forgot, there's a second instance of the fallback stuff that lives in Precompilation, could you apply any of the patterns you used to fix the functionals to the precompilaton module (in a different PR of course)

https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNet.Mvc/RazorPreCompileModule.cs#L63

dougbu commented:
hmm, somewhat different scenario there.  suggest you file another bug similar to #1612 in the MVC repo with your thoughts on the importance / urgency of a fix.

HaoK commented:
Well #1612 already talks about the razor pre compilation fallback hack, so we can just keep that bug alive to track removal of that instance

dougbu commented:
@HaoK updated FYI.  will push later this evening unless you object to the fix for the
>  second instance of the fallback stuff that lives in Precompilation

issue [discussed](https://github.com/aspnet/Mvc/pull/2137#issuecomment-77799795) above.

commits are rebased and mostly squashed.  new stuff should be limited to c1270ce and 8408987.

dougbu commented:
@HaoK updated. I cleaned up the delegating `IServiceProvider` implementation but also moved the pre-compilation `ApplicationBasePath` workarounds into `RazorPreCompileModule` (not just one of its sub-classes).  noticed the other sites were not doing pre-compilation during functional test runs.

can't find any remaining `IServiceProvider` implementations in MVC.

@Eilon please chime in if you think the `ApplicationBasePath` override has no place in product code.  from my perspective, duplicating this code in three web sites was ugly and didn't align with the work @ianhong is doing to make the framework easier to test.

HaoK commented:
Looks good!  Finally all the old fallback hacks are all gone :)

davidfowl commented:
Almost :smile:

dougbu commented:
@davidfowl updated per "choice 2 (fallback)" in our offline discussion.  would have needed to override the `RazorViewEngineOptions` setup to avoid `IApplicationEnvironment` entirely.  suspect customers are more likely to override the  `RazorViewEngineOptions` setup.

davidfowl commented:
Thats ok for now. Will bug @pranavkm when he gets back to fix it all properly.

dougbu commented on samples/TagHelperSample.Web/Compiler/Preprocess/RazorPreCompilation.cs:
```diff
@@ -9,13 +9,11 @@ namespace TagHelperSample.Web
-                                            "debug",
-                                            StringComparison.OrdinalIgnoreCase);
+            GenerateSymbols =
+                string.Equals(applicationEnvironment.Configuration, "debug", StringComparison.OrdinalIgnoreCase);
```
will revert whitespace changes in this file when I squash

dougbu commented on samples/MvcSample.Web/Compiler/Preprocess/RazorPreCompilation.cs:
```diff
@@ -9,13 +9,11 @@ namespace MvcSample.Web
-                                            "debug",
-                                            StringComparison.OrdinalIgnoreCase);
+            GenerateSymbols =
+                string.Equals(applicationEnvironment.Configuration, "debug", StringComparison.OrdinalIgnoreCase);
```
will revert whitespace changes in this file when I squash

davidfowl commented:
:shipit:

dougbu commented:
commit cc7b319
